pyscriptjs/tests/support.py

```diff
@@ -0,0 +1,319 @@
+        )
+
+    def pyscript_run(self, snippet):
+        """
+        Main entry point for pyscript tests.
+
+        snippet contains a fragment of HTML which will be put inside a full
+        HTML document. In particular, the <head> automatically contains the
+        correct <script> and <link> tags which are necessary to load pyscript
+        correctly.
+
+        This method does the following:
+          - write a full HTML file containing the snippet
+          - open a playwright page for it
+          - wait until pyscript has been fully loaded
+        """
+        doc = f"""
+        <html>
+          <head>
+              <link rel="stylesheet" href="{self.http_server}/build/pyscript.css" />
+              <script defer src="{self.http_server}/build/pyscript.js"></script>
+          </head>
+          <body>
+            {snippet}
+          </body>
+        </html>
+        """
+        filename = f"{self.testname}.html"
+        self.writefile(filename, doc)
+        self.goto(filename)
+        self.wait_for_pyscript()
+
+
+# ============== Helpers and utility functions ==============
+
+
+class JsError(Exception):
+    """
+    Represent an exception which happened in JS.
+
+    It's a thin wrapper around playwright.sync_api.Error, with two important
+    differences:
+
+    1. it has a better name: if you see JsError in a trabeback, it's
+       immediately obvious that it's a JS exception.
+
+    2. Show also the JS stacktrace by default, contrarily to
+       playwright.sync_api.Error
+    """
+
+    def __init__(self, error):
+        super().__init__(self.format_playwright_error(error))
+        self.error = error
+
+    @staticmethod
+    def format_playwright_error(error):
+        # apparently, playwright Error.stack contains all the info that we
+        # want: exception name, message and stacktrace. The docs say that
+        # error.stack is optional, so fallback to the standard repr if it's
+        # unavailable.
+        return error.stack or str(error)
+
+
+class JsMultipleErrors(Exception):
+    """
+    This is raised in case we get multiple JS errors in the page
+    """
+
+    def __init__(self, errors):
+        lines = ["Multiple JS errors found:"]
+        for err in errors:
+            lines.append(JsError.format_playwright_error(err))
+        msg = "\n".join(lines)
+        super().__init__(msg)
+        self.errors = errors
+
+
+class ConsoleMessageCollection:
+    """
+    Helper class to collect and expose ConsoleMessage in a Pythonic way.
+
+    Usage:
+
+      console.log.messages: list of ConsoleMessage with type=='log'
+      console.log.lines:    list of strings
+      console.log.text:     the whole text as single string
+
+      console.debug.*       same as above, but with different types
+      console.info.*
+      console.error.*
+      console.warning.*
+
+      console.all.*         same as above, but considering all messages, no filters
+    """
+
+    class View:
```

Comment:
Any reason for not defining it at the global scope and then passing it explicitly as class attribute?

Reply:
no big reason. It's just that this class is used/usable only inside `ConsoleMessageCollection`, so I didn't want to pollute the global scope unnecessarily.